### ISSUE TITLE
Unable to complete round trip tests for a FLAC File using libsnd library

### ISSUE THREAD

**Issue by upanitag** [libsndfile/libsndfile#504]
Hi Libsnd community,

I am trying to write some data to a .FLAC file and read it back using libsnd. Unfortunately, for certain values the data written to the file and the data read from the file is not equal. For example, writing the value 0.999938964843750 to the flac file gives back the same result during read. But writing a value of 0.999999999999833 does not yield the same value at read. Writing -1 also does not work correctly when reading it back.

This issue was found when writing a normalized sine wave to a .FLAC file and reading back the data in MATLAB. The read output was attenuated. In order to confirm the issue, I wrote a simple program and picked a few random values from the sine wave. 

See code below:

#include "sndfile.h"
#include <cstdio.h>
#include <assert.h>
#include <iostream.h>
// ------------------------------------------------
// libsndfile virtual file I/O function pointers
// ------------------------------------------------
/** Return current position of the virtual file context
 * @param   user_data   user defined file context (used for FILE pointer in this header)
 * @return  sf_count_t  virtual file's current position in byte
 */
static sf_count_t tellVirtualIO (void *user_data)
{
    FILE* pFile = reinterpret_cast<FILE*>(user_data);

    fpos_t fpos;
    if ( fgetpos(pFile, &fpos) )
    {                    
        return (sf_count_t) -1; //fgetpos() returns non-zero on failure
    }

    #ifdef __linux__
        return static_cast<sf_count_t>(fpos.__pos);
    #else
        return static_cast<sf_count_t>(fpos);
    #endif
}

/** Seek virtual file to offset
 * @param   offset      offset from whence to seek to
 * @param   whence      seek reference point. Defined in libsndfile as one of SEEK_CUR, SEEK_SET or SEEK_END
 * @param   user_data   user defined file context (used for FILE pointer in this header)
 * @return  sf_count_t  virtual file's new offset in byte
 */
static sf_count_t seekVirtualIO (sf_count_t offset, int whence, void *user_data)
{               
    FILE* pFile = reinterpret_cast<FILE*>(user_data);

    // libsndfile demands whence equals one of these
    assert( whence==SEEK_CUR || whence==SEEK_SET || whence==SEEK_END );

    // call the platform's 64-bit ready fseek()
    #if defined (_WIN32)
        _fseeki64(pFile, offset, whence);
    #elif defined(__APPLE__)
        fseeko(pFile, offset, whence);
    #else // (UNIX)
        fseeko64(pFile, offset, whence);
    #endif
                
    if ( ferror(pFile) )
    {
        return (sf_count_t) -1; 
    }

    return tellVirtualIO(user_data);
}

/** Get length of the virtual file in bytes
 * @param   user_data   user defined file context (used for FILE pointer in this header)
 * @return  sf_count_t  virtual file's length in bytes
 */
static sf_count_t getFileLenVirtualIO (void *user_data)
{
    seekVirtualIO(0, SEEK_END, user_data); //SEEK_END defined in sndfile.h
    sf_count_t filelen = tellVirtualIO(user_data);                
    rewind( reinterpret_cast<FILE*>(user_data) );
    return filelen;
}

/** Copy "count" bytes into data buffer provided by "user_data"
 * @param   ptr         pointer to block of memory with a size of at least (size*count) bytes
 * @param   count       number of bytes to be copied to buffer
 * @param   user_data   user defined file context (used for FILE pointer in this header)
 * @return  sf_count_t  number of bytes actually copied to buffer
 */
static sf_count_t readVirtualIO (void *ptr, sf_count_t count, void *user_data)
{
    FILE* pFile = reinterpret_cast<FILE*>(user_data);
    size_t bytesRead = fread(ptr, sizeof(uint8_t), static_cast<size_t>(count), pFile);

    if ( ferror(pFile) ) // ferror() equals zero if fread suceeded
    {
        return (sf_count_t) 0;
    }

    return static_cast<sf_count_t>(bytesRead);
}

/** Process "count" bytes stored in the buffer provided by "user_data"
 * @param   ptr         pointer to array of elements to be written
 * @param   count       number of bytes to be processed from buffer
 * @param   user_data   user defined file context (used for FILE pointer in this header)
 * @return  sf_count_t  number of bytes actually processed
 */
static sf_count_t writeVirtualIO (const void *ptr, sf_count_t count, void *user_data)
{
    FILE* pFile = reinterpret_cast<FILE*>(user_data);
    size_t bytesWritten = fwrite(ptr, sizeof(uint8_t), static_cast<size_t>(count), pFile);

    if (ferror(pFile) != 0) // fwrite failed
    {
        return (sf_count_t) 0;
    }

    return static_cast<sf_count_t>(bytesWritten);
}
int main()
{
    SF_INFO fileInfo;
    SF_VIRTUAL_IO vio;
    SNDFILE* sndfile;
    FILE* file;
    fileInfo.frames = 0;
    fileInfo.samplerate = 44100;
    fileInfo.channels = 1;
    fileInfo.format = 1507330;
    fileInfo.sections = 0;
    fileInfo.seekable = 0;
    
    file = fopen("mytest.flac", "wb");

    vio.get_filelen = getFileLenVirtualIO; // virtualIO function ptr: return file length in bytes
    vio.seek = seekVirtualIO; // virtualIO function ptr: seek to offset and return new location in bytes
    vio.read = readVirtualIO; // virtualIO function ptr: copy data to buffer and return processed in bytes
    vio.write = writeVirtualIO; // virtualIO function ptr: copy data from buffer and return processed in bytes
    vio.tell = tellVirtualIO; // virtualIO function ptr: return current position of virtual file context in bytes

    //const double data[] = {0.539554165872951,0.999999999999833, 0.908558587800332};
    const double data[] = {
          0.999938964843750,
          0.999969482421875,
          0.998901367187500,
          0.539554165872951,
          0.999999999999833,
          0.999869482421876 ,
          0.999938964843750, 
          0.998901367187500, 
          0.999999999999833, 
          0.5,
          1.0,
          -1.000000000000000
    };

    // Open the file for writing
    sndfile = sf_open_virtual(&vio, SFM_WRITE, &fileInfo, reinterpret_cast<void*>(file)); 
    sf_command(sndfile, SFC_SET_SCALE_INT_FLOAT_WRITE, NULL, SF_TRUE);
    sf_command(sndfile, SFC_SET_CLIPPING, NULL, SF_TRUE);
    sf_count_t framesWritten = sf_writef_double(sndfile, &data[0], 12);
    sf_close(sndfile);
    fclose(file);

    // Read the data - Round trip test
    file = fopen("mytest.flac","r");
    sndfile = sf_open_virtual(&vio, SFM_READ, &fileInfo, reinterpret_cast<void*>(file));
    sf_count_t startSample = 0;
    sf_count_t totalframesToRead = 12;
    sf_count_t framesSeeked = sf_seek(sndfile, startSample, SEEK_SET );

    double* buffer = (double*)malloc(fileInfo.channels  * totalframesToRead *sizeof(double));
    sf_count_t framesRead = sf_readf_double(sndfile, &buffer[0], totalframesToRead);
    for(int i = 0 ; i < framesRead; ++i)
    {
        std::cout << "Input frame data::" << data[i] << "....."<< "Ouput frame data " <<  buffer[i] << std::endl;
        if(data[i] == buffer[i])
            std::cout <<"frameNos " << i+1 <<  " completed successful roundtrip operation" << std::endl;
        else
            std::cout <<"frameNos " << i+1 <<  " did not complete successful roundtrip operation" << std::endl;
    }
    free(buffer);
    buffer = nullptr;
    sf_close(sndfile);
    fclose(file);
}
/*********************************END-OF-CODE************************************/

I am using libsnd 1.0.28 to reproduce the problem in Linux.

1. CPU Info 
processor	: 7
vendor_id	: GenuineIntel
cpu family	: 6
model		: 85
model name	: Intel(R) Xeon(R) W-2133 CPU @ 3.60GHz

2. Memory Info
MemTotal:       65621876 kB
MemFree:         2720732 kB
MemAvailable:   31587344 kB

3. Operating System : Debian-9

4. libsnd version 1.0.28

**Comment by upanitag** [comment]
#include "sndfile.h"
#include <cstdio>
#include <assert.h>

#include <iostream>

// ------------------------------------------------
// libsndfile virtual file I/O function pointers
// ------------------------------------------------
/** Return current position of the virtual file context
 * @param   user_data   user defined file context (used for FILE pointer in this header)
 * @return  sf_count_t  virtual file's current position in byte
 */
static sf_count_t tellVirtualIO (void *user_data)
{
    FILE* pFile = reinterpret_cast<FILE*>(user_data);

    fpos_t fpos;
    if ( fgetpos(pFile, &fpos) )
    {                    
        return (sf_count_t) -1; //fgetpos() returns non-zero on failure
    }

    #ifdef __linux__
        return static_cast<sf_count_t>(fpos.__pos);
    #else
        return static_cast<sf_count_t>(fpos);
    #endif
}

/** Seek virtual file to offset
 * @param   offset      offset from whence to seek to
 * @param   whence      seek reference point. Defined in libsndfile as one of SEEK_CUR, SEEK_SET or SEEK_END
 * @param   user_data   user defined file context (used for FILE pointer in this header)
 * @return  sf_count_t  virtual file's new offset in byte
 */
static sf_count_t seekVirtualIO (sf_count_t offset, int whence, void *user_data)
{               
    FILE* pFile = reinterpret_cast<FILE*>(user_data);

    // libsndfile demands whence equals one of these
    assert( whence==SEEK_CUR || whence==SEEK_SET || whence==SEEK_END );

    // call the platform's 64-bit ready fseek()
    #if defined (_WIN32)
        _fseeki64(pFile, offset, whence);
    #elif defined(__APPLE__)
        fseeko(pFile, offset, whence);
    #else // (UNIX)
        fseeko64(pFile, offset, whence);
    #endif
                
    if ( ferror(pFile) )
    {
        return (sf_count_t) -1; 
    }

    return tellVirtualIO(user_data);
}

/** Get length of the virtual file in bytes
 * @param   user_data   user defined file context (used for FILE pointer in this header)
 * @return  sf_count_t  virtual file's length in bytes
 */
static sf_count_t getFileLenVirtualIO (void *user_data)
{
    seekVirtualIO(0, SEEK_END, user_data); //SEEK_END defined in sndfile.h
    sf_count_t filelen = tellVirtualIO(user_data);                
    rewind( reinterpret_cast<FILE*>(user_data) );
    return filelen;
}

/** Copy "count" bytes into data buffer provided by "user_data"
 * @param   ptr         pointer to block of memory with a size of at least (size*count) bytes
 * @param   count       number of bytes to be copied to buffer
 * @param   user_data   user defined file context (used for FILE pointer in this header)
 * @return  sf_count_t  number of bytes actually copied to buffer
 */
static sf_count_t readVirtualIO (void *ptr, sf_count_t count, void *user_data)
{
    FILE* pFile = reinterpret_cast<FILE*>(user_data);
    size_t bytesRead = fread(ptr, sizeof(uint8_t), static_cast<size_t>(count), pFile);

    if ( ferror(pFile) ) // ferror() equals zero if fread suceeded
    {
        return (sf_count_t) 0;
    }

    return static_cast<sf_count_t>(bytesRead);
}

/** Process "count" bytes stored in the buffer provided by "user_data"
 * @param   ptr         pointer to array of elements to be written
 * @param   count       number of bytes to be processed from buffer
 * @param   user_data   user defined file context (used for FILE pointer in this header)
 * @return  sf_count_t  number of bytes actually processed
 */
static sf_count_t writeVirtualIO (const void *ptr, sf_count_t count, void *user_data)
{
    FILE* pFile = reinterpret_cast<FILE*>(user_data);
    size_t bytesWritten = fwrite(ptr, sizeof(uint8_t), static_cast<size_t>(count), pFile);

    if (ferror(pFile) != 0) // fwrite failed
    {
        return (sf_count_t) 0;
    }

    return static_cast<sf_count_t>(bytesWritten);
}
int main()
{
    SF_INFO fileInfo;
    SF_VIRTUAL_IO vio;
    SNDFILE* sndfile;
    FILE* file;
    fileInfo.frames = 0;
    fileInfo.samplerate = 44100;
    fileInfo.channels = 1;
    fileInfo.format = 1507330;
    fileInfo.sections = 0;
    fileInfo.seekable = 0;
    
    file = fopen("mytest.flac", "wb");

    vio.get_filelen = getFileLenVirtualIO; // virtualIO function ptr: return file length in bytes
    vio.seek = seekVirtualIO; // virtualIO function ptr: seek to offset and return new location in bytes
    vio.read = readVirtualIO; // virtualIO function ptr: copy data to buffer and return processed in bytes
    vio.write = writeVirtualIO; // virtualIO function ptr: copy data from buffer and return processed in bytes
    vio.tell = tellVirtualIO; // virtualIO function ptr: return current position of virtual file context in bytes

    //const double data[] = {0.539554165872951,0.999999999999833, 0.908558587800332};
    const double data[] = {
          0.999938964843750,
          0.999969482421875,
          0.998901367187500,
          0.539554165872951,
          0.999999999999833,
          0.999869482421876 ,
          0.999938964843750, 
          0.998901367187500, 
          0.999999999999833, 
          0.5,
          1.0,
          -1.000000000000000
    };

    // Open the file for writing
    sndfile = sf_open_virtual(&vio, SFM_WRITE, &fileInfo, reinterpret_cast<void*>(file)); 
    sf_command(sndfile, SFC_SET_SCALE_INT_FLOAT_WRITE, NULL, SF_TRUE);
    sf_command(sndfile, SFC_SET_CLIPPING, NULL, SF_TRUE);
    sf_count_t framesWritten = sf_writef_double(sndfile, &data[0], 12);
    sf_close(sndfile);
    fclose(file);

    // Read the data - Round trip test
    file = fopen("mytest.flac","r");
    sndfile = sf_open_virtual(&vio, SFM_READ, &fileInfo, reinterpret_cast<void*>(file));
    sf_count_t startSample = 0;
    sf_count_t totalframesToRead = 12;
    sf_count_t framesSeeked = sf_seek(sndfile, startSample, SEEK_SET );

    double* buffer = (double*)malloc(fileInfo.channels  * totalframesToRead *sizeof(double));
    sf_count_t framesRead = sf_readf_double(sndfile, &buffer[0], totalframesToRead);
    for(int i = 0 ; i < framesRead; ++i)
    {
        std::cout << "Input frame data::" << data[i] << "....."<< "Ouput frame data " <<  buffer[i] << std::endl;
        if(data[i] == buffer[i])
            std::cout <<"frameNos " << i+1 <<  " completed successful roundtrip operation" << std::endl;
        else
            std::cout <<"frameNos " << i+1 <<  " did not complete successful roundtrip operation" << std::endl;
    }
    free(buffer);
    buffer = nullptr;
    sf_close(sndfile);
    fclose(file);
}

**Comment by upanitag** [comment]
#include "sndfile.h"
#include <cstdio>
#include <assert.h>
#include <iostream>

// ------------------------------------------------
// libsndfile virtual file I/O function pointers
// ------------------------------------------------
/** Return current position of the virtual file context
 * @param   user_data   user defined file context (used for FILE pointer in this header)
 * @return  sf_count_t  virtual file's current position in byte
 */
static sf_count_t tellVirtualIO (void *user_data)
{
    FILE* pFile = reinterpret_cast<FILE*>(user_data);

    fpos_t fpos;
    if ( fgetpos(pFile, &fpos) )
    {                    
        return (sf_count_t) -1; //fgetpos() returns non-zero on failure
    }

    #ifdef __linux__
        return static_cast<sf_count_t>(fpos.__pos);
    #else
        return static_cast<sf_count_t>(fpos);
    #endif
}

/** Seek virtual file to offset
 * @param   offset      offset from whence to seek to
 * @param   whence      seek reference point. Defined in libsndfile as one of SEEK_CUR, SEEK_SET or SEEK_END
 * @param   user_data   user defined file context (used for FILE pointer in this header)
 * @return  sf_count_t  virtual file's new offset in byte
 */
static sf_count_t seekVirtualIO (sf_count_t offset, int whence, void *user_data)
{               
    FILE* pFile = reinterpret_cast<FILE*>(user_data);

    // libsndfile demands whence equals one of these
    assert( whence==SEEK_CUR || whence==SEEK_SET || whence==SEEK_END );

    // call the platform's 64-bit ready fseek()
    #if defined (_WIN32)
        _fseeki64(pFile, offset, whence);
    #elif defined(__APPLE__)
        fseeko(pFile, offset, whence);
    #else // (UNIX)
        fseeko64(pFile, offset, whence);
    #endif
                
    if ( ferror(pFile) )
    {
        return (sf_count_t) -1; 
    }

    return tellVirtualIO(user_data);
}

/** Get length of the virtual file in bytes
 * @param   user_data   user defined file context (used for FILE pointer in this header)
 * @return  sf_count_t  virtual file's length in bytes
 */
static sf_count_t getFileLenVirtualIO (void *user_data)
{
    seekVirtualIO(0, SEEK_END, user_data); //SEEK_END defined in sndfile.h
    sf_count_t filelen = tellVirtualIO(user_data);                
    rewind( reinterpret_cast<FILE*>(user_data) );
    return filelen;
}

/** Copy "count" bytes into data buffer provided by "user_data"
 * @param   ptr         pointer to block of memory with a size of at least (size*count) bytes
 * @param   count       number of bytes to be copied to buffer
 * @param   user_data   user defined file context (used for FILE pointer in this header)
 * @return  sf_count_t  number of bytes actually copied to buffer
 */
static sf_count_t readVirtualIO (void *ptr, sf_count_t count, void *user_data)
{
    FILE* pFile = reinterpret_cast<FILE*>(user_data);
    size_t bytesRead = fread(ptr, sizeof(uint8_t), static_cast<size_t>(count), pFile);

    if ( ferror(pFile) ) // ferror() equals zero if fread suceeded
    {
        return (sf_count_t) 0;
    }

    return static_cast<sf_count_t>(bytesRead);
}

/** Process "count" bytes stored in the buffer provided by "user_data"
 * @param   ptr         pointer to array of elements to be written
 * @param   count       number of bytes to be processed from buffer
 * @param   user_data   user defined file context (used for FILE pointer in this header)
 * @return  sf_count_t  number of bytes actually processed
 */
static sf_count_t writeVirtualIO (const void *ptr, sf_count_t count, void *user_data)
{
    FILE* pFile = reinterpret_cast<FILE*>(user_data);
    size_t bytesWritten = fwrite(ptr, sizeof(uint8_t), static_cast<size_t>(count), pFile);

    if (ferror(pFile) != 0) // fwrite failed
    {
        return (sf_count_t) 0;
    }

    return static_cast<sf_count_t>(bytesWritten);
}
int main()
{
    SF_INFO fileInfo;
    SF_VIRTUAL_IO vio;
    SNDFILE* sndfile;
    FILE* file;
    fileInfo.frames = 0;
    fileInfo.samplerate = 44100;
    fileInfo.channels = 1;
    fileInfo.format = 1507330;
    fileInfo.sections = 0;
    fileInfo.seekable = 0;
    
    file = fopen("mytest.flac", "wb");

    vio.get_filelen = getFileLenVirtualIO; // virtualIO function ptr: return file length in bytes
    vio.seek = seekVirtualIO; // virtualIO function ptr: seek to offset and return new location in bytes
    vio.read = readVirtualIO; // virtualIO function ptr: copy data to buffer and return processed in bytes
    vio.write = writeVirtualIO; // virtualIO function ptr: copy data from buffer and return processed in bytes
    vio.tell = tellVirtualIO; // virtualIO function ptr: return current position of virtual file context in bytes

    //const double data[] = {0.539554165872951,0.999999999999833, 0.908558587800332};
    const double data[] = {
          0.999938964843750,
          0.999969482421875,
          0.998901367187500,
          0.539554165872951,
          0.999999999999833,
          0.999869482421876 ,
          0.999938964843750, 
          0.998901367187500, 
          0.999999999999833, 
          0.5,
          1.0,
          -1.000000000000000
    };

    // Open the file for writing
    sndfile = sf_open_virtual(&vio, SFM_WRITE, &fileInfo, reinterpret_cast<void*>(file)); 
    sf_command(sndfile, SFC_SET_SCALE_INT_FLOAT_WRITE, NULL, SF_TRUE);
    sf_command(sndfile, SFC_SET_CLIPPING, NULL, SF_TRUE);
    sf_count_t framesWritten = sf_writef_double(sndfile, &data[0], 12);
    sf_close(sndfile);
    fclose(file);

    // Read the data - Round trip test
    file = fopen("mytest.flac","r");
    sndfile = sf_open_virtual(&vio, SFM_READ, &fileInfo, reinterpret_cast<void*>(file));
    sf_count_t startSample = 0;
    sf_count_t totalframesToRead = 12;
    sf_count_t framesSeeked = sf_seek(sndfile, startSample, SEEK_SET );

    double* buffer = (double*)malloc(fileInfo.channels  * totalframesToRead *sizeof(double));
    sf_count_t framesRead = sf_readf_double(sndfile, &buffer[0], totalframesToRead);
    for(int i = 0 ; i < framesRead; ++i)
    {
        std::cout << "Input frame data::" << data[i] << "....."<< "Ouput frame data " <<  buffer[i] << std::endl;
        if(data[i] == buffer[i])
            std::cout <<"frameNos " << i+1 <<  " completed successful roundtrip operation" << std::endl;
        else
            std::cout <<"frameNos " << i+1 <<  " did not complete successful roundtrip operation" << std::endl;
    }
    free(buffer);
    buffer = nullptr;
    sf_close(sndfile);
    fclose(file);
}

**Comment by erikd** [comment]
This is a know issue when reading/writing normalized float data to a sound file that stores that data as integers. It will not be fixed. See http://www.mega-nerd.com/libsndfile/FAQ.html#Q010

**Comment by upanitag** [comment]
> This is a know issue when reading/writing normalized float data to a sound file that stores that data as integers. It will not be fixed. See http://www.mega-nerd.com/libsndfile/FAQ.html#Q010


Thanks for the reply. I am suprised that writing -1.000000000000000 returned 1 in the code which I sent you. Also, what is the cause of the attenuation when writing normalized sine wave  ( See jpeg file attached).

![attenuated](https://user-images.githubusercontent.com/47428547/71734657-619e1980-2e1a-11ea-9a81-0b803fae223c.jpg)

**Comment by upanitag** [comment]
zoomed in picture
![zoomedin](https://user-images.githubusercontent.com/47428547/71734832-d2453600-2e1a-11ea-91a5-d8dcaa720b4d.jpg)

**Comment by cwannaz** [comment]
I can reproduce what @upanitag describes above. Interestingly, this is triggered (indirectly I suppose) by the presence of -1 values in the signal.

Running the code below without correction leads to the following:
![image](https://user-images.githubusercontent.com/96522/71742662-7edde280-2e30-11ea-9f87-dc98c8de1bb2.png)

where the first plot represents the signal written to the flac file, the second the signal read from the flac file, the third a flag "normalized signal == -1", and the last the difference between the signals written/read to/from the file.

Running the code below with correction leads to the following:
![image](https://user-images.githubusercontent.com/96522/71742760-c95f5f00-2e30-11ea-892c-aab3d64b7a7a.png)

In both cases, the signal is a sine wave (pretty rough given the ratio f/fs). The only difference is that the correction bumps up all values of the normalized signal strictly equal to -1, from a value equal to the numeric precision at that level (~2.2204e-16).

Repro code:
`clear
close all

fs = 1e3 ;
f  = fs / 11.2 ;

signal = sin( 2*pi * f/fs * (0:1e3) ) ;
signal_norm = signal / max( abs( signal )) ;

if false % correction
    signal_norm(signal_norm <= -1) = -1 + eps(-1) ;
    color = 'r' ;
else
    color = 'b' ;
end

locator = 'signal_norm.flac' ;
audiowrite( locator, signal_norm, fs ) ;
pause( 0.1 ) ;
signal_read = audioread( locator ).' ;

figure ;
subplot( 4, 1, 1 ) ;
plot( signal_norm, color ) ;
title( 'Normalized signal' ) ;

subplot( 4, 1, 2 ) ;
plot( signal_read, color ) ;
title( 'Signal read from flac file' ) ;

subplot( 4, 1, 3 ) ;
plot( abs(signal_norm + 1) == 0, color ) ;
title( 'Flag "signal\_norm==-1"' ) ;

subplot( 4, 1, 4 ) ;
delta = signal_norm - signal_read ;
plot( delta, color ) ;
title( sprintf( 'Difference - max abs diff = %.2g', max( abs( delta )))) ;`

**Comment by erikd** [comment]
> Thanks for the reply. I am surprised that writing -1.000000000000000 returned 1 in the code which I sent you.

Oh I missed that point. Yes, that i definitely wrong.

**Comment by F-Haferkorn** [comment]
Hallo Erik de-Castro-Lopo,
Hallo to  the erik/libsndfile Newsgroup-Team,

My name is Frank Haferkorn am a Physicist specialized (besides many other) in Algorithmic_s and Signal processing.
I am listening to this mailgroup for a long time
I am using lbsndfile for my own projects,  And I am very please with it.
On the long term, I would like to participate in your project...

THIS THREAD caught my interest.


  1.  Does this issue also appear when
     *   Writing with to a float data file Format
     *   in opposite to the (16bit) signed integer  fiile  format?
  2.  I  did not understand when/where the normalization happened
  3.  SAQs: (Seldom asked questions)
     *   Do you get a different behaviour at higher or lower sample rates.
     *   This looks to me as a kind of “impulse-answer” as a reply of the irregular value.
  4.  Is the WAV-data and the analysis results somewhere available on the internet?
     *   I would like to take a look at that data and thee analysis scripts (gnuplot is OK) soo far.
     *   Can you notify me, where to find the related files on the net?
  5.  How many users of libsndfile have this problem?

I see another problem...

  1.  In my eyes this conversion<->reconversion problem
will block away any scientific users that rely to reliable numerical precise results.
Do you see this, too???

Dipl.-Phys. Frank Haferkorn



Von: upanitag<mailto:notifications@github.com>
Gesendet: Freitag, 3. Januar 2020 17:17
An: erikd/libsndfile<mailto:libsndfile@noreply.github.com>
Cc: Subscribed<mailto:subscribed@noreply.github.com>
Betreff: Re: [erikd/libsndfile] Unable to complete round trip tests for a FLAC File using libsnd library (#504)


This is a know issue when reading/writing normalized float data to a sound file that stores that data as integers. It will not be fixed. See http://www.mega-nerd.com/libsndfile/FAQ.html#Q010

Thanks for the reply. I am suprised that writing -1.000000000000000 returned 1 in the code which I sent you. Also, what is the cause of the attenuation when writing normalized sine wave ( See jpeg file attached).

[attenuated]<https://user-images.githubusercontent.com/47428547/71734657-619e1980-2e1a-11ea-9a81-0b803fae223c.jpg>

**Comment by erikd** [comment]
> 1.  Does this issue also appear when
>     *   Writing with to a float data file Format
>     *   in opposite to the (16bit) signed integer  fiile  format?

The FLAC format does not support float data format.

This issue *only* arises when the data provided by the calling program is floating point and the data format of the file is integer and it is independent of the destination file format.

>   2.  I  did not understand when/where the normalization happened

The normalization happens within libsndfile itself.

>  3.  SAQs: (Seldom asked questions)
>     *   Do you get a different behaviour at higher or lower sample rates.

No, the same problem exists regardless of sample rate.

> 1.  In my eyes this conversion<->reconversion problem
> will block away any scientific users that rely to reliable numerical precise results.

libsndfile is designed for audio data. If anyone chooses to use it for scientific data they should first evaluate libsndfile to see if it is suitable for their needs.

**Comment by F-Haferkorn** [comment]
[:###:]
Hallo Erik. De-Castro-Lopo,,,,

Thank you for the quick and friendly reply.
3. Do you get a different behaviour at higher or lower sample rates.

>No, the same problem exists regardless of sample rate.
(a1)
GREAT, one problem less.



  1.  In my eyes this conversion<->reconversion problem
will block away any scientific users that rely to reliable numerical precise results.

> libsndfile is designed for audio data.
> If anyone chooses to use it for scientific data they should first evaluate libsndfile to see if it is suitable for their needs.

(a2)
OK, see me as an experienced SW-expert && as a scientist

I am not introduced into the netiquette of this newsgroup....
(q1)       Is it OK if I have a look at the code on this issue.?
(q2)       Where can I find the related “round trip tests”

Yours,,,,
                Dipl.-Phys, Frank Haferkorn, Ottobrunn

**Comment by upanitag** [comment]
> > Thanks for the reply. I am surprised that writing -1.000000000000000 returned 1 in the code which I sent you.
> 
> Oh I missed that point. Yes, that i definitely wrong and needs to be fixed.

Thansk Eric. Looking forward to the fix.

**Comment by F-Haferkorn** [comment]
[:###:]                  // had to resent this mail due to a Problem with my mail-tool
Hallo Erik. De-Castro-Lopo,,,,

Thank you for the quick and friendly reply.
3. Do you get a different behaviour at higher or lower sample rates.

>No, the same problem exists regardless of sample rate.
(a1): GREAT, one problem less.

  1.  In my eyes this conversion<->reconversion problem
will block away any scientific users that rely to reliable numerical precise results.

> libsndfile is designed for audio data.
> If anyone chooses to use it for scientific data they should first evaluate libsndfile to see if it is suitable for their needs.

(a2):      OK, see me as an experienced SW-expert && as a scientist

I am not introduced into the netiquette of this newsgroup....
(q1):      Is it OK if I have a look at the code on this issue.?
(q2):       Where can I find the related “round trip tests”

Yours,,,,
                Dipl.-Phys, Frank Haferkorn, Ottobrunn

**Comment by erikd** [comment]
@F-Haferkorn This is not a newsgroup nor is it an email list. Its is a github issue : https://github.com/erikd/libsndfile/issues/504

**Comment by gbeckers** [comment]
Just wondering if there is progress on this issue. I just got bitten by it.

**Comment by upanitag** [comment]
I was wondering if there is any update/fix for this issue?

**Comment by evpobr** [comment]
Hello everybody. Sorry for this bug, let's try to fix it. @roebel, pull request?

**Comment by evpobr** [comment]
@upanitag, here is output of your program and libsndfile 1.0.31 on Windows platform:

```
D:\source\repos\sndfile-flac-bug\build>"D:\source\repos\sndfile-flac-bug\build\sndfile-flac-bug.exe"
Input frame data::0.999939.....Ouput frame data 0.999939  
frameNos 1 completed successful roundtrip operation       
Input frame data::0.999969.....Ouput frame data 0.999969  
frameNos 2 completed successful roundtrip operation       
Input frame data::0.998901.....Ouput frame data 0.998901  
frameNos 3 completed successful roundtrip operation       
Input frame data::0.539554.....Ouput frame data 0.539551  
frameNos 4 did not complete successful roundtrip operation
Input frame data::1.....Ouput frame data 0.999969
frameNos 5 did not complete successful roundtrip operation
Input frame data::0.999869.....Ouput frame data 0.999878
frameNos 6 did not complete successful roundtrip operation
Input frame data::0.999939.....Ouput frame data 0.999939
frameNos 7 completed successful roundtrip operation
Input frame data::0.998901.....Ouput frame data 0.998901
frameNos 8 completed successful roundtrip operation
Input frame data::1.....Ouput frame data 0.999969
frameNos 9 did not complete successful roundtrip operation
Input frame data::0.5.....Ouput frame data 0.5
frameNos 10 completed successful roundtrip operation
Input frame data::1.....Ouput frame data 0.999969
frameNos 11 did not complete successful roundtrip operation
Input frame data::-1.....Ouput frame data -1
frameNos 12 completed successful roundtrip operation
```

**Comment by evpobr** [comment]
Ubuntu 20.04.2 LTS with libsndfile 1.0.28 is buggy:

```
Input frame data::0.999939.....Ouput frame data 0.999939
frameNos 1 completed successful roundtrip operation
Input frame data::0.999969.....Ouput frame data 0.999969
frameNos 2 completed successful roundtrip operation
Input frame data::0.998901.....Ouput frame data 0.998901
frameNos 3 completed successful roundtrip operation
Input frame data::0.539554.....Ouput frame data 0.539551
frameNos 4 did not complete successful roundtrip operation
Input frame data::1.....Ouput frame data 0.999969
frameNos 5 did not complete successful roundtrip operation
Input frame data::0.999869.....Ouput frame data 0.999878
frameNos 6 did not complete successful roundtrip operation
Input frame data::0.999939.....Ouput frame data 0.999939
frameNos 7 completed successful roundtrip operation
Input frame data::0.998901.....Ouput frame data 0.998901
frameNos 8 completed successful roundtrip operation
Input frame data::1.....Ouput frame data 0.999969
frameNos 9 did not complete successful roundtrip operation
Input frame data::0.5.....Ouput frame data 0.5
frameNos 10 completed successful roundtrip operation
Input frame data::1.....Ouput frame data 0.999969
frameNos 11 did not complete successful roundtrip operation
Input frame data::-1.....Ouput frame data 1
frameNos 12 did not complete successful roundtrip operation
```

with output to WAV works:

```
Input frame data::0.999939.....Ouput frame data 0.999939
frameNos 1 completed successful roundtrip operation
Input frame data::0.999969.....Ouput frame data 0.999969
frameNos 2 completed successful roundtrip operation
Input frame data::0.998901.....Ouput frame data 0.998901
frameNos 3 completed successful roundtrip operation
Input frame data::0.539554.....Ouput frame data 0.539551
frameNos 4 did not complete successful roundtrip operation
Input frame data::1.....Ouput frame data 0.999969
frameNos 5 did not complete successful roundtrip operation
Input frame data::0.999869.....Ouput frame data 0.999847
frameNos 6 did not complete successful roundtrip operation
Input frame data::0.999939.....Ouput frame data 0.999939
frameNos 7 completed successful roundtrip operation
Input frame data::0.998901.....Ouput frame data 0.998901
frameNos 8 completed successful roundtrip operation
Input frame data::1.....Ouput frame data 0.999969
frameNos 9 did not complete successful roundtrip operation
Input frame data::0.5.....Ouput frame data 0.5
frameNos 10 completed successful roundtrip operation
Input frame data::1.....Ouput frame data 0.999969
frameNos 11 did not complete successful roundtrip operation
Input frame data::-1.....Ouput frame data -1
frameNos 12 completed successful roundtrip operation
```

**Comment by evpobr** [comment]
@upanitag, can you check again with libsndfile 1.0.31?

**Comment by krisfed** [comment]
I was just trying it out, and it seems the issue was already resolved as early as 1.0.30 :)

**Comment by evpobr** [comment]
Ok, thanks for info.
